src/kernel/boot.c

```diff
@@ -873,21 +881,64 @@ BOOT_CODE bool_t init_freemem(word_t n_available, const p_region_t *available,
      * from. If possible, retain any left over memory as an extra p region */
     word_t size = calculate_rootserver_size(it_v_reg, extra_bi_size_bits);
     word_t max = rootserver_max_size_bits(extra_bi_size_bits);
+    pptr_t start;
     for (; i >= 0; i--) {
-        word_t next = i + 1;
-        pptr_t start = ROUND_DOWN(ndks_boot.freemem[i].end - size, max);
-        if (start >= ndks_boot.freemem[i].start) {
-            create_rootserver_objects(start, it_v_reg, extra_bi_size_bits);
-            if (i < ARRAY_SIZE(ndks_boot.freemem)) {
-                ndks_boot.freemem[next].end = ndks_boot.freemem[i].end;
-                ndks_boot.freemem[next].start = start + size;
+        // SimplExportAnRefine has poor handling of array indices that are sums
+        // of variables and small constants. The `next_i` variable exists to
+        // work around the deficiency.
+        word_t next_i = i + 1;
+        region_t *reg = &ndks_boot.freemem[i];
+        /* Check if the region is big enough. If so, align the start address
+         * down to ensure the required alignment and then check that the start
+         * address is still within the region.
+         */
+        if (reg->end - reg->start >= size) {
+            start = ROUND_DOWN(reg->end - size, max);
+            if (start >= reg->start) {
+                break;
             }
-            ndks_boot.freemem[i].end = start;
-            break;
-        } else if (i < ARRAY_SIZE(ndks_boot.freemem)) {
-            ndks_boot.freemem[next] = ndks_boot.freemem[i];
         }
+        /* The region is too small. Since we loop though the list from the end,
+         * and we know there is an empty region at the end, we can be sure that
+         * if we arrive here, there is an empty region in the next higher slot.
+         * Swap the slots and try the next lower region:
+         *
+         *                region too small
+         *                v
+         * Free: [a] [b] [c] [empty] [...]  ->  [a] [b] [empty] [c] [...]
+         *            ^
+         *            region to check next
+         */
+        region_t *next = &ndks_boot.freemem[next_i];
```

Comment:
Maybe we can avoid the need for having `next_i` completely by just using:
```
region_t *next = reg + 1;
```

Reply:
Might be best to read my other comments before investing any more effort in my flawed commit. :-)